### PR TITLE
Add GFA diff operation

### PR DESCRIPTION
**Pull request by dkhofer**

### CODE REVIEW

File: src/diffs/gfa.rs [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+                sequence: source_sequence[start..end].to_string(),
+                node_id: source_block.node_id,
+                sequence_start: source_block.sequence_start,
+                strand: source_block.strand,
+            };
+            segments.insert(segment.clone());
+            source_segment_ids.push(segment.segment_id());
+            source_strands.push(source_block.strand);
+        }
+        for (block1, block2) in source_node_blocks.iter().tuple_windows() {
```
Comment (Chris7) [inline src/diffs/gfa.rs:103]: just a heads up that tuple_windows will be effectively a noop on a list of length 1. Will the uses here always have >= 2 items?

File: src/diffs/gfa.rs [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+use crate::models::{block_group::BlockGroup, path::Path, sample::Sample};
+use crate::range::Range;
+
+pub fn gfa_sample_diff(
+    conn: &Connection,
+    collection_name: &str,
+    filename: &PathBuf,
+    from_sample_name: Option<&str>,
+    to_sample_name: &str,
+) {
```
Comment (Chris7) [inline src/diffs/gfa.rs:18]: can you put in a comment the general approach here?

File: src/diffs/gfa.rs [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+        if mappings.is_empty() {
+            continue;
+        }
+        let mut source_ranges = vec![];
+        let mut target_ranges = vec![];
+
+        let mut source_mapping_start = 0;
+        let mut target_mapping_start = 0;
+        for mapping in &mappings {
+            if mapping.source_range.start > source_mapping_start {
```
Comment (Chris7) [inline src/diffs/gfa.rs:53]: can you put in a comment about the source_mapping_start and what is happening here? I'm confusing why we have this.

File: src/diffs/gfa.rs (hunk 1 above)
Comment (Chris7) [reply src/diffs/gfa.rs:103]: looking at node_block_parition, looks like this could be a single block.

File: src/models/path.rs [hunk 4]
```diff
@@ -235,29 +242,12 @@ impl Path {
         sequences_by_node_id: &HashMap<i64, Sequence>,
         current_path_length: i64,
     ) -> PathBlock {
-        if into.target_node_id != out_of.source_node_id {
```
Comment (dkhofer) [inline src/models/path.rs:238]: These checks aren't needed here any more because of the validations I added

File: src/diffs/gfa.rs [hunk 5]
```diff
@@ -0,0 +1,644 @@
...
+    let target_paths_by_name = target_block_groups
+        .iter()
+        .map(|bg| (bg.name.clone(), BlockGroup::get_current_path(conn, bg.id)))
+        .collect::<HashMap<String, Path>>();
+
+    let mut segments = HashSet::new();
+    let mut links = HashSet::new();
+    let mut paths = vec![];
+
+    for (target_name, target_path) in target_paths_by_name.iter() {
```
Comment (Chris7) [inline src/diffs/gfa.rs:38]: if a path gets deleted, wouldn't it not be present here?

File: src/diffs/gfa.rs [hunk 6]
```diff
@@ -0,0 +1,644 @@
...
+            source_ranges.push(mapping.source_range.clone());
+            source_mapping_start = mapping.source_range.end;
+            if mapping.target_range.start > target_mapping_start {
+                target_ranges.push(Range {
+                    start: target_mapping_start,
+                    end: mapping.target_range.start,
+                });
+            }
+            target_ranges.push(mapping.target_range.clone());
+            target_mapping_start = mapping.target_range.end;
```
Comment (Chris7) [inline src/diffs/gfa.rs:68]: can we call this something like last_position or something? would make it more obvious what this is in how its used below.

File: src/diffs/gfa.rs (hunk 2 above)
Comment (dkhofer) [reply src/diffs/gfa.rs:22]: Done

File: src/diffs/gfa.rs (hunk 1 above)
Comment (dkhofer) [reply src/diffs/gfa.rs]: It should be fine.  If there is a single block then there is only one segment, and we want to not generate any links in that case

Comment (dkhofer) [reply src/diffs/gfa.rs]: But I appreciate the warning!

File: src/diffs/gfa.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (dkhofer) [reply src/diffs/gfa.rs]: I reworked the logic to handle the case where the source or the target has a block group that the other lacks

File: src/diffs/gfa.rs (hunk 6 above)
Comment (dkhofer) [reply src/diffs/gfa.rs]: Yep, renamed to last_source_position and last_target_position, makes sense

File: src/diffs/gfa.rs [hunk 7]
```diff
@@ -0,0 +1,1058 @@
...
+                phased: 0,
+            })
+            .collect::<Vec<BlockGroupEdgeData>>();
+        BlockGroupEdge::bulk_create(&conn, &block_group_edges);
+
+        let _path1 = Path::create(&conn, "parent", block_group.id, &edge_ids);
+
+        let temp_dir = tempdir().unwrap();
+        let gfa_path = temp_dir.path().join("self-diff.gfa");
+        gfa_sample_diff(&conn, collection_name, &gfa_path, None, "child");
```
Comment (Chris7) [inline src/diffs/gfa.rs:542]: i'm confused why this would work when the `child` sample doesn't exist.

Comment (dkhofer) [reply src/diffs/gfa.rs]: I misnamed this test.  It was actually comparing a sample with stuff against the None sample, which didn't have anything (no block groups).  So it's like doing a git diff when you've just created a file and it prints out the whole file.

That said, I also just added a test comparing a sample to itself...

File: src/diffs/gfa.rs [hunk 8]
```diff
@@ -0,0 +1,1058 @@
...
+
+        import_gfa(&gfa_path, "test collection 2", None, &conn);
```
[the remark this answers is not in the document]
Comment (dkhofer) [reply src/diffs/gfa.rs:643]: I'm...not sure, at least for GFA format.  In general, we provide a GFA with all segments, links, and paths, not just the segments that are different.  My preference is to be consistent with that style even for a self diff.

I do want to put together a PR that prints out a pure sequence diff on the command line, closer to what git does, and I agree in that case, the output should be empty

File: src/diffs/gfa.rs [hunk 9]
```diff
@@ -0,0 +1,1157 @@
...
+            }
+            target_ranges.push(mapping.target_range.clone());
+            last_target_position = mapping.target_range.end;
+        }
+
+        if has_source_path {
+            let source_path = source_path_result.unwrap();
+            let source_sequence = source_path.sequence(conn);
+
+            let source_len = source_sequence.len().try_into().unwrap();
```
Comment (Chris7) [inline src/diffs/gfa.rs:120]: i think you can do this like `xxx.len() as i64`

File: src/diffs/gfa.rs [hunk 10]
```diff
@@ -0,0 +1,1157 @@
...
+        writer
+            .write_all(&path_line(&path).into_bytes())
+            .unwrap_or_else(|_| panic!("Error writing path {} to GFA stream", path.name));
+    }
+}
+
+fn segments_from_blocks(node_blocks: &Vec<NodeIntervalBlock>, sequence: &str) -> Vec<Segment> {
+    let mut segments = vec![];
+    for block in node_blocks {
+        let start = usize::try_from(block.start).unwrap();
```
Comment (Chris7) [inline src/diffs/gfa.rs:180]: `block.start as usize` might be more concise here too

File: src/diffs/gfa.rs [hunk 11]
```diff
@@ -0,0 +1,1157 @@
...
+            target_strand: block2.strand,
+        };
+        links.push(link);
+    }
+
+    links
+}
+
+fn path_from_segments(sample_name: Option<&str>, path: &Path, segments: &[Segment]) -> GFAPath {
+    let path_name = if sample_name.is_some() && sample_name.unwrap() != "" {
```
Comment (Chris7) [inline src/diffs/gfa.rs:227]: maybe do `sample_name.unwrap_or("") != ""`?

File: src/diffs/gfa.rs [hunk 12]
```diff
@@ -0,0 +1,1157 @@
...
+                "AACCCCAATTGGGGTT",
+                "AAAAAAAATTGGGGTT"
+            ]
+            .iter()
+            .map(|s| s.to_string())
+            .collect::<HashSet<String>>()
+        );
+
+        let gfa_path = temp_dir.path().join("child-grandchild-diff.gfa");
+        gfa_sample_diff(
```
Comment (Chris7) [inline src/diffs/gfa.rs:454]: :clap:

File: src/diffs/gfa.rs [hunk 13]
```diff
@@ -0,0 +1,1157 @@
...
+            ["AACCCCAATTTTTTTT", "AACCCCAATTGGGGTT"]
+                .iter()
+                .map(|s| s.to_string())
+                .collect::<HashSet<String>>()
+        );
+    }
+
+    #[test]
+    fn test_gfa_diff_against_nothing() {
+        // Confirm diff of a sample to itself is just a single segment
```
Comment (Chris7) [inline src/diffs/gfa.rs:481]: `sample to nothing`

File: src/diffs/gfa.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Chris7) [reply src/diffs/gfa.rs:643]: gotcha. I think we'll probably need to refine how it's called. If I do `diff a.txt a.txt`, I expect nothing. I understand why the GFA format is different. It's focused on the edges/etc. between things.

File: src/diffs/gfa.rs [hunk 14]
```diff
@@ -0,0 +1,1157 @@
...
+            ["AAAAAAAATTTTTTTT"]
+                .iter()
+                .map(|s| s.to_string())
+                .collect::<HashSet<String>>()
+        );
+    }
+
+    #[test]
+    fn test_gfa_diff_unrelated_paths() {
+        // Confirm diff of a sample to itself is just a single segment
```
Comment (Chris7) [inline src/diffs/gfa.rs:661]: update test comment info

Comment (Chris7) [reply src/diffs/gfa.rs:661]: there's a few tests w/ this same info copy/pasted

**Review by Chris7 — APPROVED** [review]
Minor stuff. LGTM! Much easier to process once I considered it as a path diff and the goal being a GFA showing deltas.

File: src/diffs/gfa.rs (hunk 9 above)
Comment (dkhofer) [reply src/diffs/gfa.rs]: Nice, done

File: src/diffs/gfa.rs (hunk 10 above)
Comment (dkhofer) [reply src/diffs/gfa.rs]: Yes.  Done

File: src/diffs/gfa.rs (hunk 11 above)
Comment (dkhofer) [reply src/diffs/gfa.rs]: Yep, done

File: src/diffs/gfa.rs (hunk 13 above)
Comment (dkhofer) [reply src/diffs/gfa.rs]: Fixed

File: src/diffs/gfa.rs [hunk 15]
```diff
@@ -0,0 +1,1157 @@
...
+                phased: 0,
+            })
+            .collect::<Vec<BlockGroupEdgeData>>();
+        BlockGroupEdge::bulk_create(&conn, &block_group_edges);
+
+        let _path1 = Path::create(&conn, "test path", block_group.id, &edge_ids);
+
+        let temp_dir = tempdir().unwrap();
+        let gfa_path = temp_dir.path().join("diff-against-nothing.gfa");
+        gfa_sample_diff(&conn, collection_name, &gfa_path, None, "test sample");
```
[the remark this answers is not in the document]
Comment (dkhofer) [reply src/diffs/gfa.rs]: I had been thinking in terms of "source sample, target sample" but it kind of doesn't matter, so I just changed it so they're both Option<&str>

File: src/diffs/gfa.rs (hunk 14 above)
Comment (dkhofer) [reply src/diffs/gfa.rs]: Oops, fixed

**Pull request merged**